internal/gatewayapi/translator.go

```diff
@@ -1721,91 +1565,11 @@ func (t *Translator) ProcessUDPRoutes(udpRoutes []*v1alpha2.UDPRoute, gateways [
-
-			if backendRef.Group != nil && *backendRef.Group != "" {
-				parentRef.SetCondition(udpRoute,
-					v1beta1.RouteConditionResolvedRefs,
-					metav1.ConditionFalse,
-					v1beta1.RouteReasonInvalidKind,
-					"Group is invalid, only the core API group (specified by omitting the group field or setting it to an empty string) is supported",
-				)
-				continue
-			}
-
-			if backendRef.Kind != nil && *backendRef.Kind != KindService {
-				parentRef.SetCondition(udpRoute,
-					v1beta1.RouteConditionResolvedRefs,
-					metav1.ConditionFalse,
-					v1beta1.RouteReasonInvalidKind,
-					"Kind is invalid, only Service is supported",
-				)
-				continue
-			}
-
-			if backendRef.Namespace != nil && string(*backendRef.Namespace) != "" && string(*backendRef.Namespace) != udpRoute.Namespace {
-				if !isValidCrossNamespaceRef(
-					crossNamespaceFrom{
-						group:     v1beta1.GroupName,
-						kind:      KindUDPRoute,
-						namespace: udpRoute.Namespace,
-					},
-					crossNamespaceTo{
-						group:     "",
-						kind:      KindService,
-						namespace: string(*backendRef.Namespace),
-						name:      string(backendRef.Name),
-					},
-					resources.ReferenceGrants,
-				) {
-					parentRef.SetCondition(udpRoute,
-						v1beta1.RouteConditionResolvedRefs,
-						metav1.ConditionFalse,
-						v1beta1.RouteReasonRefNotPermitted,
-						fmt.Sprintf("Backend ref to service %s/%s not permitted by any ReferenceGrant", *backendRef.Namespace, backendRef.Name),
-					)
-					continue
-				}
-			}
-
-			if backendRef.Port == nil {
-				parentRef.SetCondition(udpRoute,
-					v1beta1.RouteConditionResolvedRefs,
-					metav1.ConditionFalse,
-					"PortNotSpecified",
-					"A valid port number corresponding to a port on the Service must be specified",
-				)
-				continue
-			}
-
 			// TODO: [v1alpha2-v1beta1] Replace with NamespaceDerefOr when UDPRoute graduates to v1beta1.
 			serviceNamespace := NamespaceDerefOrAlpha(backendRef.Namespace, udpRoute.Namespace)
 			service := resources.GetService(serviceNamespace, string(backendRef.Name))
-			if service == nil {
-				parentRef.SetCondition(udpRoute,
-					v1beta1.RouteConditionResolvedRefs,
-					metav1.ConditionFalse,
-					v1beta1.RouteReasonBackendNotFound,
-					fmt.Sprintf("Service %s/%s not found", serviceNamespace, string(backendRef.Name)),
-				)
-				continue
-			}
-
-			var portFound bool
-			for _, port := range service.Spec.Ports {
-				if port.Port == int32(*backendRef.Port) && port.Protocol == v1.ProtocolUDP {
-					portFound = true
-					break
-				}
-			}
 
-			if !portFound {
-				parentRef.SetCondition(udpRoute,
-					v1beta1.RouteConditionResolvedRefs,
-					metav1.ConditionFalse,
-					"PortNotFound",
-					fmt.Sprintf("UDP Port %d not found on service %s/%s", *backendRef.Port, serviceNamespace,
-						string(backendRef.Name)),
-				)
+			if !checkBackendRef(&backendRef, parentRef, udpRoute, resources, serviceNamespace, KindUDPRoute) {
 				continue
```

Comment:
need a warn log, a metric will be better

Reply:
A condition will be set to the route if the backendRef can't pass validation.  Furthermore, if we want to add log to translator, we need to do it in a consistent way, which means we may want to add log to all the checks, not only here. It would be better in another PR if we think this is necessary.

Reply:
reasonable, got it.